### ISSUE TITLE
npm install mangles package.json in node_modules, breaking verification

### ISSUE THREAD

**Issue by hach-que** [RedpointGames/pkgsign#4]
When `npm install` runs, it updates `package.json` of modules to include metadata about the registry it was obtained from. 

**Workaround**

`yarn install` does not do this, so the workaround for those impacted by this issue is to use `yarn` instead until a fix is made.

**Details**

As an example, this is what `npm install` mangles `package.json` to for `pkgsign` itself:

```json
{
  "_args": [
    [
      "pkgsign@0.1.2",
      "C:\\Users\\June\\Documents\\Projects\\pkgsign-test"
    ]
  ],
  "_from": "pkgsign@0.1.2",
  "_id": "pkgsign@0.1.2",
  "_inBundle": false,
  "_integrity": "sha1-ulN8bCht7GtmOI/tOrphGn3c8es=",
  "_location": "/pkgsign",
  "_phantomChildren": {},
  "_requested": {
    "type": "version",
    "registry": true,
    "raw": "pkgsign@0.1.2",
    "name": "pkgsign",
    "escapedName": "pkgsign",
    "rawSpec": "0.1.2",
    "saveSpec": null,
    "fetchSpec": "0.1.2"
  },
  "_requiredBy": [
    "/"
  ],
  "_resolved": "https://registry.npmjs.org/pkgsign/-/pkgsign-0.1.2.tgz",
  "_spec": "0.1.2",
  "_where": "C:\\Users\\June\\Documents\\Projects\\pkgsign-test",
  "author": {
    "name": "Redpoint Games"
  },
  "bin": {
    "pkgsign": "dist/index.js"
  },
  "bugs": {
    "url": "https://github.com/RedpointGames/pkgsign/issues"
  },
  "dependencies": {
    "clime": "^0.5.9",
    "eol-fix-stream": "^1.0.0",
    "isbinaryfile": "^3.0.2",
    "node-cmd": "^3.0.0",
    "node-fetch": "^1.7.3",
    "npm-packlist": "^1.1.10",
    "openpgp": "^2.6.1",
    "prompt": "^1.0.0",
    "recursive-readdir": "^2.2.1",
    "targz": "^1.0.1",
    "tmp": "^0.0.33"
  },
  "description": "A CLI tool for signing and verifying npm and yarn packages.",
  "devDependencies": {
    "@types/node": "^8.5.7",
    "@types/node-fetch": "^1.6.7",
    "@types/recursive-readdir": "^2.2.0",
    "@types/tmp": "^0.0.33",
    "rimraf": "^2.6.2",
    "ts-node": "^3.3.0",
    "typescript": "^2.6.1"
  },
  "homepage": "https://github.com/RedpointGames/pkgsign#readme",
  "license": "MIT",
  "main": "index.js",
  "name": "pkgsign",
  "private": false,
  "repository": {
    "type": "git",
    "url": "git+https://github.com/RedpointGames/pkgsign.git"
  },
  "scripts": {
    "build": "rimraf dist && tsc",
    "pkgsign": "rimraf dist && tsc && node dist/index.js",
    "sign": "rimraf dist && tsc && node dist/index.js sign .",
    "signedpack": "rimraf dist && tsc && yarn pack --filename pkgsign.tgz && node dist/index.js sign pkgsign.tgz",
    "signedpublish": "rimraf dist && tsc && yarn pack --filename pkgsign.tgz && node dist/index.js sign pkgsign.tgz && yarn publish pkgsign.tgz",
    "verify": "rimraf dist && tsc && node dist/index.js verify . --full"
  },
  "version": "0.1.2"
}
```

Not only is the order of properties modified (`version` isn't the last element in the original `package.json`), but we never included any of those `_` properties, and they don't exist in the tarball uploaded to `npm` (otherwise `yarn install` would also display this issue, as it would be unable to recover the original `package.json` data).

**Comment by hach-que** [comment]
I'm most likely going to target some sort of fix for this after 0.2.0, when I bring in a new `signature.json` format designed to be extensible. It's quite hard to fix as we need to parse `package.json` and compare the tree of values to ensure there aren't any modified values, new or missing values, with the exception of `_` prefixed top-level properties.

This also means we'd need to exclude `package.json` from the `files` signature, which I'm not particularly happy about.

**Comment by simonepri** [comment]
Hello @hach-que,
Maybe I've not understood the problem entirely, but does not seem too hard to me.

Sign:
1) Strip all `_` prefixed top-level properties of the `package.json`
2) Sort all the keys lexicographically
2) Compute the signature for the `package.json`

Verify:
1) Strip all `_` prefixed top-level properties of the `package.json`
2) Sort all the keys lexicographically
2) Verify the signature

The only problem I see, is that stripping and sorting all `packages.json` increase the computational complexity of the verify process.

See [this](https://www.npmjs.com/package/removeNPMAbsolutePaths) for the stripping process.

**Comment by hach-que** [comment]
Yeah, it's not that much more complex, though it is a special-casing of file signing which is currently generic for every file in the package.

It concerns me more that NPM or Node.js could add some semantic runtime meaning to an underscore prefixed property, and our verification process would ignore it, allowing an exploit. I would much rather we had a way to tell NPM to not mangle `project.json`, so that the file is what the original author signed. Because our verification process runs after `postinstall` scripts, you could have one dependency inject `_` prefixed properties into another dependency, and verification would not pick it up. Or you could have packages already which read `_` properties from `project.json` for metadata about themselves!

I'm probably going to submit a request to NPM first, see if I can get some changes made there, and if not use the technique to exclude `_` metadata (maybe with an explicit whitelist of names and expected value types instead of just a prefix check).

**Comment by hach-que** [comment]
This is now resolved in `0.2.0-alpha.4`, which can be installed with `npm i -g pkgsign@next`, `yarn global add pkgsign@next` or by installing from the Git repository's `dev` branch as per instructions in the README.
